jcl/src/java.base/share/classes/java/lang/ClassLoader.java

```diff
@@ -1413,6 +1413,7 @@ private static boolean registerAsParallelCapable(Class<?> callerCls) {
 
 	/*[IF JAVA_SPEC_VERSION >= 19]*/
 	if (!ClassLoader.class.isAssignableFrom(superCls)) {
+		// check the superclass of the caller to only allow subclasses of ClassLoader
 		throw new IllegalCallerException(callerCls.getName());
 	}
 	/*[ENDIF] JAVA_SPEC_VERSION >= 19 */
```

Comment:
I think this check should be done earlier. If the classloader is already registered, a call to this method will succeed no matter who the caller is, contrary to the documentation: https://docs.oracle.com/en/java/javase/18/docs/api/java.base/java/lang/ClassLoader.html#registerAsParallelCapable()

Reply:
I've updated to move the check first to handle a null caller.